build.sbt

```diff
@@ -519,6 +522,105 @@ lazy val componentModulesPaths =
+  val fullCp    = (runnerCp ++ runtimeCp).distinct
+  val log       = streams.value.log
+  JPMSUtils.filterModulesFromClasspath(
+    fullCp,
+    GraalVM.modules,
+    log,
+    shouldContainAll = true
+  )
+}
+
+/** The javaOptions are created such that the `runtime/test` will run with truffle-compiler
+  * So we need to properly create --module-path option.
+  * Module `org.enso.runtime` is added to --module-path as a path to a directory with
+  * expanded classes, i.e., not a Jar archive. This means we also have to properly create
+  * --patch-module option. This is an optimization so that we don't have to assembly the
+  * `runtime.jar` fat jar for the tests.
+  */
+lazy val modulePathTestOptions =
+  taskKey[Seq[String]](
+    "Provides javaOptions for running tests with --module-path set such that " +
+    "enso (`org.enso.runtime`) module and truffle-compiler and the rest of " +
+    "the GraalVM modules are on module-path. Also makes sure that the `runtime.jar` " +
+    "fat jar is not rebuilt unnecessarily"
+  )
+(ThisBuild / modulePathTestOptions) := {
+  val updateReport = (LocalProject("runtime") / Test / update).value
+  val runtimeModName = (LocalProject(
+    "runtime-with-instruments"
+  ) / moduleInfos).value.head.moduleName
+  val runtimeMod = (LocalProject(
+    "runtime-with-instruments"
+  ) / Compile / productDirectories).value
+  val graalMods = graalModulesPaths.value
+  val loggingMods = JPMSUtils.filterModulesFromUpdate(
+    updateReport,
+    logbackPkg ++ Seq(
+      "org.slf4j" % "slf4j-api" % slf4jVersion
+    ),
+    streams.value.log,
+    shouldContainAll = true
+  )
+
+  /** All these modules will be in --patch-module cmdline option to java, which means that
+    * for the JVM, it will appear that all the classes contained in these sbt projects are contained
+    * in the `org.enso.runtime` module. In this way, we do not have to assembly the `runtime.jar`
+    * fat jar.
+    */
+  val modulesToPatchIntoRuntime: Seq[File] =
+    (LocalProject(
+      "runtime-instrument-common"
+    ) / Compile / productDirectories).value ++
+    (LocalProject(
+      "runtime-instrument-id-execution"
+    ) / Compile / productDirectories).value ++
+    (LocalProject(
+      "runtime-instrument-repl-debugger"
+    ) / Compile / productDirectories).value ++
+    (LocalProject(
+      "runtime-instrument-runtime-server"
+    ) / Compile / productDirectories).value ++
+    (LocalProject(
+      "runtime-language-epb"
+    ) / Compile / productDirectories).value ++
+    (LocalProject("runtime-compiler") / Compile / productDirectories).value
+  val patchStr = modulesToPatchIntoRuntime
+    .map(_.getAbsolutePath)
+    .mkString(File.pathSeparator)
+  val allModulesPaths: Seq[String] =
+    runtimeMod.map(_.getAbsolutePath) ++
+    graalMods.map(_.data.getAbsolutePath) ++
+    loggingMods.map(_.getAbsolutePath)
+  // We can't use org.enso.logger.TestLogProvider (or anything from our own logging framework here) because it is not
```

Comment:
That would mean it will render our configuration useless?

Reply:
Useless is a very harsh word in general. It just cannot be used for testing, if we want to start the forked JVM with `--module-path`. The problem is [split packages](https://stackoverflow.com/a/51828213/4816269), i.e., we have for example `org.enso.logging` package distributed in many projects, and if we would like to produce a single Jar module per single project, we cannot succeed.

Do you think it is problematic to use just plain `ch.qos.logback` for the test logging?

Reply:
> Do you think it is problematic to use just plain ch.qos.logback for the test logging?

Well, I think _useless_ is the right word in this case because there is no point of `TestLogProvider` anymore, unless I'm missing something. I'm not sure the setup will work now as expected, though. Can you double check that the output from `sbt> test` is now comparable?

Reply:
At the end, I am using `TestLogProvider` - a342562c2dd98b2ab5c1bd22e1ccb8ba82f82582. So there should be no difference in the test output with develop.